### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by bobi-no**
Update for implementation of reserved $error variable.

### CODE REVIEW

**Review by petrowsky — CHANGES_REQUESTED** [review]
Let's get a proper header block on this function.

File: functions/Errors/Error.fmfn
Comment (petrowsky) [inline functions/Errors/Error.fmfn:16]: We need to update ErrorEnv to ErrorEnvironment.

File: functions/Errors/Error.fmfn
Comment (petrowsky) [inline functions/Errors/Error.fmfn:23]: We should remove this note entirely if direct calls are being made to Error as opposed to calling ErrorApp or ErrorFmp which are simply convenience functions.

File: functions/Errors/ErrorEnvironment.fmfn
Comment (petrowsky) [inline functions/Errors/ErrorEnvironment.fmfn:1]: We need to update ErrorEnv to ErrorEnvironment at lines 12 & 13

File: functions/Errors/ErrorFound.fmfn
Comment (petrowsky) [inline functions/Errors/ErrorFound.fmfn:1]: Given that we're reserving the $error variable for all error handling, we could probably update this to test for empty errorData and then default to looking at $error.

Line 38, and beyond, could be like this.
```
Let ( [
	~errorData = If ( IsEmpty ( errorData ) ; $error ; errorData );

	~jsonCode = JSONGetElement ( ~errorData ; "code" );
	~jsonType = JSONGetElementType ( ~errorData ; "code" )
];
	Case ( 
		IsEmpty (~errorData ) ;
			False; 
	
		Exact ( ~errorData ; Filter ( ~errorData ; "-0123456789" ) ) ;
			GetAsBoolean ( ~errorData );
	
		~jsonType = JSONNumber ;
			GetAsBoolean ( ~jsonCode );
	
		Left ( ~jsonType ; 1 ) = "?" ;
			GetAsBoolean ( ~jsonCode );
	
		/*default result*/ 
			True
	)
)
```

This would allow for a script step line calling **If [ ErrorFound("") ]** instead of having to pass in **If [ ErrorFound($error) ]**

File: functions/Errors/ErrorLog.fmfn
Comment (petrowsky) [inline functions/Errors/ErrorLog.fmfn:1]: We need a proper header on this function and turn it into a general wrapper. Meaning we would remove the specific call to MBS and provide a note - allowing the end developer to opt to use other plug-ins.

We can keep the reference to the MBS plug-in as a dev note for developers who may not know they can create records using SQL.

If you want me to adjust this one I can. Just let me know.

File: functions/Errors/Error.fmfn
Comment (bobi-no) [reply functions/Errors/Error.fmfn:23]: Done.

File: functions/Errors/ErrorLog.fmfn
Comment (bobi-no) [reply functions/Errors/ErrorLog.fmfn:1]: As far as I am concerned, ErrorLog does not need to be part of the function set. I see error handling and error logging as somewhat separate. I'm not sure this framework should favor a specific logging implementation, and I feel like including this function binds us to do so.

That said, if the function stays in this collection, could we keep in a comment a sample call using MBS? Removing the plugin function call seems like it will gut the function quite extensively in this case and even impact the test we have on ~result = "OK". (Perhaps I need to see what your end result would look like)

If there is a way to convey the proper idea where the dev has a clue of what to backfill this whole with, then I guess we can go this way, but generally speaking I would rather provide one way that works and let people adapt as they need, VS giving a generic blueprint in a non-working state and the dev will wonder if they broke it, if what they went with matches the original or not, etc.

Comment (bobi-no) [reply functions/Errors/ErrorLog.fmfn:1]: If you can adjust this one as you see fit, I would appreciate.

Comment (dansmith65) [reply functions/Errors/ErrorLog.fmfn:1]: > I see error handling and error logging as somewhat separate.

I wholeheartedly agree with this statement! In fact I would say they are completely separate concerns; sure, I usually want to log error's, but I also want to log other stuff and I don't log all errors (User Canceled Action, for example).

File: functions/Errors/ErrorFound.fmfn
Comment (bobi-no) [reply functions/Errors/ErrorFound.fmfn:1]: I understand the convenience for the call to be **If [ ErrorFound("") ]**. That said, I'm not convinced I like how that call looks. Without defaulting to $error when the param is empty, **ErrorFound($error)**  would likely be the predominant call seen throughout a system, with other possible calls here and there.

With the default in place, **ErrorFound("")** would likely be predominant. To a reader who is not familiar with the functions, **ErrorFound("")** can remain ambiguous as to what it does and how it does it. Something that is not the case with the **ErrorFound($error)** version of the function call.

Also, the function itself is used by `ErrorInScriptResult(  )`, and I would have to test how it behaves in that context (I assume it would fallback on a $error telling us the script result is not valid JSON, but it would remain to be tested.

Finally, I find there is some use in having $error exposed in other calcs and scripts. I know some devs like to make custom functions that will simply return a reserved variable, and even if I understand that it provides completion as you type, avoids typos, inconsistencies and lets you rename the reserved variable if you so desire, I feel like it obfuscates what makes the whole thing work, to the point where someone may actually overwrite $error by mistake because the code is not giving ample cues that the variable is reserved.

If a dev is bothered by the function not falling back on $error when no other input is provided, they can easily remedy that, but I remain shy to make that the de-facto implementation.

Comment (dansmith65) [reply functions/Errors/ErrorFound.fmfn:1]: > To a reader who is not familiar with the functions, `ErrorFound("")` can remain ambiguous as to what it does and how it does it.

Unfortunately, when using a reserved variable in any of these functions, it will always be ambiguous to users not familiar with the system. That said, it's a trade-off I've made for some of the functions I use right now as well, mostly spurred by @jwillinghalpern. The primary reason for the reserved variable is code readability. I can use a text expander to "type" a repetitive snippet easily enough, but long error trapping code is distracting and makes it hard to see the "real" code. This comment is meant to be applied more generally to this suite of error functions rather than to this one function in particular, though.

You could add yet another function: `ErrorFoundIn ( $error )`. Then drop the parameter entirely from the primary CF, so it would appear as `ErrorFound()` or `ErrorFound`. Otherwise, I agree that you might as well pass the parameter in.

Comment (jwillinghalpern) [reply functions/Errors/ErrorFound.fmfn:1]: I've thought about adding a parameter-less companion to `ErrorFound($error)`: `ErrorExists`.

```
If [ ErrorExists ] 
  // handle $error
Else
```

But the main appeal for me is less about the length and more about providing a way to check for errors set by functions that store a default `$error` var. It always feels weird to mix and match:

```
Set Field []
# function that checks for error and stores in $error magically
Exit Loop If [ ErrorCheckLast("Set Field") ]

# later :(, who set $error??
If [ ErrorFound($error) ]

# maybe it's better if not naming the var when setting it, to not name it when checking it too:
If [ ErrorExists() ]
```

Comment (bobi-no) [reply functions/Errors/ErrorFound.fmfn:1]: Yes, readability is a strong motivation for this compromise. I do find that `ErrorFound("")` harms readability, more than `ErrorApp ( 1 ; "User cancellation." )`even if both of them do not reveal their use of `$error`.

Also, even if I like the idea of adding that feature in the function while introducing a dedicated function like `ErrorFoundIn ( $myError )`, the set of functions as it stands does not strike me as self-explanatory. You kind of need to know what function to use where and when, what is mostly private vs public, etc. I'm afraid introducing layers of additional functions that mostly serve a utility/convenience purpose will make the set of functions seem bulkier and more complicated, and as a result will hinder adoption.

At this point, it is more about the philosophy of fmpstandards and what it is meant to be. My own implementation has an extra function that tests if a given error code is '1' because user cancellation is something that will silence my dialogs (and skip logging as @dansmith65 mentioned elsewhere). That said, I do not consider that function to be part of the "core" functions that make this error handling pattern work. Are we attempting to anticipate how everyone will roll out that pattern in their own solutions and account for required adaptations, or are we aiming for the minimal amount of code where if we remove features it makes the pattern non-viable?

I know error handling is meant as a best practice, and as such it is not as prescriptive as coding standards, but that seems like a good reason not to include too many tools to our swiss army knife. Simply saying "This is the base model. You can add or swap tools as you see fit". We clearly do not have something with only a long blade and a short blade, we are saying the 'adventure' of error handling does require a can opener and a corkscrew, but we do not feel like the magnifying lens is required to achieve results. If someone feels like they can achieve better results with that other tool, they can add it, and they can swap the corkscrew for a saw, but everyone will have different opinions about that, and baking those tools in the base model will make it look too bulky for some.

File: functions/Errors/ErrorEnvironment.fmfn
Comment (bobi-no) [reply functions/Errors/ErrorEnvironment.fmfn:1]: I changed it, but only see the change when I am editing the document. I will mark as resolved. Please re-open if the modification is not preserved.

File: functions/Errors/ErrorFound.fmfn
Comment (jwillinghalpern) [reply functions/Errors/ErrorFound.fmfn:1]: @bobi-no That's a good point about the script making clear at _some_ point that $error is reserved/set. Now I'm happy I never pulled the trigger on ErrorExists!

And in that same spirit, hiding setting $error was never a goal, the "magical" convenience wrappers were really just there to eliminate ultra-repetitive, and quite verbose boilerplate that overshadowed what Dan called the "real" code.

Thanks for that perspective, i agree.

Comment (bobi-no) [reply functions/Errors/ErrorFound.fmfn:1]: Pros & cons on both sides, really. Exposing it also means the dev is committed to the actual name of the reserved variable. Changing it after the fact would require a lot of work (relative to the footprint of the error handling implementation). I'm simply explaining why I prefer one side over the other. I tend to adopt a position that is meant to be easier on dev unfamiliar with the framework. Devs having the familiarity to opt into a different design usually have all they need to make those decisions.

File: functions/Errors/ErrorInScriptResult.fmfn
Comment (dansmith65) [inline functions/Errors/ErrorInScriptResult.fmfn:31]: I think this should use the `ErrorApp` CF instead of `Error`.

When I created the concept of "app" errors, I always intended for common errors like this to be defined in the `ErrorApp` CF itself. I'm not sure you _should_ standardize on the codes I use, but to give an example at least, here's the codes I define in my `ErrorApp` CF:

```
theErrorCode = 0 ;     "No error" ;
theErrorCode = 1 ;     "User canceled action" ;
theErrorCode = 2 ;     "Invalid Parameter" ; /* Missing or malformed, so if a date is expected but script received a turtle, parameter is invalid. */
theErrorCode = 3 ;     "Failed Validation" ; /* Any general-purpose validation that doesn't fit into another more-specific error code. */
theErrorCode = 6 ;     "Invalid window mode" ;
theErrorCode = 10 ;     "Incompatible FileMaker version" ; /* Sometimes we offer a feature or two that doesn't run on the lowest supported client version. In that case, this error could be used to only allow supported clients to run the code. An example of this is the Export All script. */
theErrorCode = 11 ;     "Incompatible FileMaker application" ; /* Might use if FMGo is required for a GPS location based function, but the feature was run from FMPro. */
theErrorCode = 13 ;     "Script Error" ; /* I often wonder if I should use this code, or Failed Validation code 3. I think there are subtle differences, though. Let's say a script set a variable to the value of "a", "b" or "c", then later tested which value it was. I would throw this code in an Else since the same script set the code, so there is just some internal error that isn't because of a parameter. It might be able to be called validation error, but the script itself is what was invalid, so I use this code.*/
theErrorCode = 16 ;     "Plug-In error" ; /* Use when a plugin doesn't have a specific error handling function (which set's the type to the plugin name) */
theErrorCode = 17 ;     "ExecuteSQL error" ;
theErrorCode = 19 ;     "Invalid script result" ; /* Set by #AssignScriptResultAndSetError script. Yes, it's basically like code 2, but for script results. You could also use it when translating a module's result to our standard, and that process fails. */

// else:
"Unknown error code"
```

You could probably replace 6, 10, and 11 with 3 (Failed Validation) but in my `ErrorApp` CF, I add relevant data to the error object for these codes: (note that this is from my old let-notation based CF)

```
Error ( "app" ; theErrorCode ; errorDescription ; theErrorInfo )
& Case (
	theErrorCode = 6 ;     # ( "WindowMode" ; Get ( WindowMode ) ) ;
	theErrorCode = 10 ;    # ( "ApplicationVersion" ; Get ( ApplicationVersion ) ) ;
	theErrorCode = 11 ;    # ( "ApplicationVersion" ; Get ( ApplicationVersion ) )
)
```

File: functions/Errors/Error.fmfn
Comment (dansmith65) [inline functions/Errors/Error.fmfn:14]: I'm not sure if this pull request is the best time to bring this up, but I'll start the conversation here and maybe we can move it elsewhere if others think that's more fitting.

Two important aspects of collecting error information for me have been:

1. Being able to specify a text description of the error, like:
    - "id_Contact parameter is required"
    - "wrong context after going to layout"
    - "Not in Find mode after going to layout."
2. Being able to add additional name/value pairs to the error to help add context (like the current value of a variable).

This function seems to combine both of those into a single `details` parameter which means you only get one or the other, but can not have both, which seems wrong to me.

Comment (bobi-no) [reply functions/Errors/Error.fmfn:14]: I believe you could put your text description in `description` and your name/value pairs as a json object in `details`

File: functions/Errors/ErrorInScriptResult.fmfn
Comment (bobi-no) [reply functions/Errors/ErrorInScriptResult.fmfn:31]: The CF was likely written before `ErrorApp (  )` existed. Calling the `Error (  )` CF directly also diminishes the CF interdependencies for devs who would implement without the `ErrorApp(  )` CF.

That said, if we see this as a minimal set of custom functions to be used together and we do have `ErrorApp` as part of the toolkit, then we should leverage it indeed.

File: functions/Errors/ErrorFmp.fmfn
Comment (bobi-no) [inline functions/Errors/ErrorFmp.fmfn:30]: Had a discussion with @dansmith65 that raised up concerns about this function call clearing the internal JSON cache even when `Get ( LastError )` = 0.

I saw value in having a $error json object with an error code of 0 along with other details for some theoretical logging situation, but I'm starting to lean the other way because of the possible performance price.

The function would then do:
```
If ( Get ( LastError ) ;
     Error ( "fmp" ; Get ( LastError ) ; "" ; details ) ;

    false)
```

The false scenario could also involve a `let(  )` statement if we want to manually build a json payload into $error.

Looking for feedback about this concept.

Comment (bobi-no) [reply functions/Errors/ErrorFmp.fmfn:30]: That pattern would be similar to the `ErrorIf (  )` pattern.

File: functions/Errors/ErrorInScriptResult.fmfn
Comment (petrowsky) [reply functions/Errors/ErrorInScriptResult.fmfn:31]: I fully agree we should limit the number of interconnections between the functions to limit dependencies. Given we're shooting for Best Practices, based on the knowledge of experienced devs, it's still important that we can support a minimal implementation with the ability to expand as desired or adapt from what's provided.

Given the above thought process, I wonder if we should revise the folder structure to the following to distinguish from the fact that ErrorApp, ErrorIf, ErrorLog, ErrorFound and ErrorScriptResult are methods which may or may not be preferred by some devs. 

- Errors
  - Core
    - Error.fmfn
    - ErrorDetails.fmfn
    - ErrorEnvironment.fmfn
    - ErrorFmp.fmfn
  - Extras
    - ErrorApp.fmfn
    - ErrorFound.fmfn
    - ErrorIf.fmfn
    - ErrorLog.fmfn
    - ErrorScriptResult.fmfn

File: functions/Errors/ErrorFmp.fmfn
Comment (petrowsky) [reply functions/Errors/ErrorFmp.fmfn:30]: FileMaker's handling of JSON is pretty bad and it's always nice to support a developer's desire to use an alternative.

File: functions/Errors/ErrorFound.fmfn
Comment (petrowsky) [reply functions/Errors/ErrorFound.fmfn:1]: Regarding the readability of the Error suite of functions, it's always a combination of both the code itself and the documentation regarding **implementation**. Once a developer reads the docs and understands the implementation, then the code becomes "more readable". Imagine trying to learn JavaScript without its documentation. You can't. And, when you see clean implementation of code, and it suits you, you tend to adopt the pattern of implementation.

Since this is the first time the standards are adopting the use of a locally scoped variable of $error, due to it's very obvious, and most likely very common, use. The documentation on the wiki should state clearly that by using the Error suite of functions the suggested implementation is to **ALWAYS** declare the $error variable at the top of a script when that script uses the Error suite. Thus, this is what is at the top of any of my scripts which handles errors using the Error suite.

```
# PURPOSE: Do something cool in FileMaker.

Allow User Abort [ Off ]
Set Error Capture [ On ]
Set Variable [ $error ] 
```

While I know that Robert is going to opt for the `Set Variable [ $error ; Value: "" ] ` (quoted version), there is absolutely no mistake that I don't know that $error is being used and it's not 'magical'. In fact, as a result of the suggested implementation, the best practice informs the developer immediately that this script has error handling incorporated. This distinction of the suggested best practice is on the implementation side and as we all know... Any hack of a developer can implement however they wish - even if it's non-readable. ;)

Comment (petrowsky) [reply functions/Errors/ErrorFound.fmfn:1]: Let's resolve this by leaving the inbound parameter as is. Any developer can expand on this and make their own supporting function without the parameter. It also leaves it possible to use the function on any other variable holding error info should such a situation actually happen.

File: functions/Errors/ErrorLog.fmfn
Comment (petrowsky) [reply functions/Errors/ErrorLog.fmfn:1]: Ok, let's simplify and kill this file. Do a git rm on your branch.

File: functions/Errors/ErrorInScriptResult.fmfn
Comment (bobi-no) [reply functions/Errors/ErrorInScriptResult.fmfn:31]: I would consider ErrorInScriptResult to be part of the core functions, but it's simply a question about where we trace the line, just like some people may see ErrorEnvironment, ErrorDetails or ErrorFMP as extras. I like the concept of labeling some as core and some as extras.

I don't want to sound picky. I'm ok with any classification we come up with.

Or we can kick this down, do a first release and do a revision once we land on where to trace the line. All of those options are fine by me.

The core/extra concept was absent in the original functions. Would we say it still applies to the original set or was the original set all considered core?

Comment (dansmith65) [reply functions/Errors/ErrorInScriptResult.fmfn:31]: > Or we can kick this down, do a first release and do a revision once we land on where to trace the line. All of those options are fine by me.

This makes sense to me. I have some pretty strong opinions on error trapping and I wouldn't mind discussing the merits of those opinions and if they should sway any code in this repo. I don't think this pull request is the best place to have that conversation, though. Maybe we can start a discussion on filemakerstandards.org?

I haven't reviewed a pull request like this in GitHub before so I've likely broken all known conventions/best practices in doing so, and I'm a little lost as to the current state of things.

Comment (bobi-no) [reply functions/Errors/ErrorInScriptResult.fmfn:31]: I don't do enough things on GitHub to be aware of conventions. When comments led to modifications I could rally to, I made the modifications. On some other things, I see the conversations as more open-ended. There are some modifications that I would rather not make in my own fork / pull request, but can be totally fine with me if it was to be altered once it got in the main branch (or a candidate branch).

To have the discussion on filemakerstandards.org, I do feel like some of the recent changes need to out of the pull request state. I don't think we want discussions there to point to a specific pull request.

The code by itself works. There could be kinks to iron out, but it's not like pushing something that does not work and is a work in progress.

It should be easier to involve others in the discussion over at fmpstandards.org

**Review by petrowsky — APPROVED** [review]

**Pull request merged**
